src/opensles/AudioStreamBuffered.cpp

```diff
@@ -25,57 +25,187 @@ namespace oboe {
+        int32_t capacity = getBufferCapacityInFrames();
+        if (capacity == oboe::kUnspecified) {
+            capacity = getFramesPerBurst() * 16; // arbitrary
+            mBufferCapacityInFrames = capacity;
+        }
+        LOGD("AudioStreamBuffered(): new FifoBuffer(bytesPerFrame=%d, capacity=%d)",
+             getBytesPerFrame(), capacity);
+        mFifoBuffer.reset(new FifoBuffer(getBytesPerFrame(), capacity));
         LOGD("AudioStreamBuffered(): mStreamCallback = %p", mStreamCallback);
     }
-    return Result::OK;
 }
 
-// TODO: This method should return a tuple of Result,int32_t where the 2nd return param is the frames written
-int32_t AudioStreamBuffered::write(const void *buffer,
-                              int32_t numFrames,
-                              int64_t timeoutNanoseconds)
-{
+int64_t AudioStreamBuffered::getFramesWritten() const {
+    if (usingFIFO()) {
+        return (int64_t) mFifoBuffer->getWriteCounter();
+    } else {
+        return AudioStream::getFramesWritten();
+    }
+}
+
+int64_t AudioStreamBuffered::getFramesRead() const {
+    if (usingFIFO()) {
+        return (int64_t) mFifoBuffer->getReadCounter();
+    } else {
+        return AudioStream::getFramesRead();
+    }
+}
+
+
+    // This is called by the OpenSL ES callback to read or write the back end of the FIFO.
+DataCallbackResult AudioStreamBuffered::onDefaultCallback(void *audioData, int numFrames) {
+    int32_t framesTransferred  = 0;
+
+    if (getDirection() == oboe::Direction::Output) {
+        // Read from the FIFO and write to audioData
+        framesTransferred = mFifoBuffer->readNow(audioData, numFrames);
+        LOGV("%s() read %d / %d frames from FIFO", __func__, framesTransferred, numFrames);
+    } else {
+        // Read from audioData and write to the FIFO
+        framesTransferred = mFifoBuffer->write(audioData, numFrames); // FIXME writeNow????
+        LOGV("%s() wrote %d / %d frames to FIFO", __func__, framesTransferred, numFrames);
+    }
+
+    if (framesTransferred < numFrames) {
+        // TODO If we do not allow FIFO to wrap then our timestamps will drift when there is an XRun!
+        incrementXRunCount();
+    }
+    markCallbackTime(numFrames); // so foreground knows how long to wait.
+    return DataCallbackResult::Continue;
+}
+
+void AudioStreamBuffered::markCallbackTime(int numFrames) {
+    mLastBackgroundSize = numFrames;
+    mBackgroundRanAtNanoseconds = AudioClock::getNanoseconds();
+}
+
+int64_t AudioStreamBuffered::predictNextCallbackTime() {
+    if (mBackgroundRanAtNanoseconds == 0) {
+        return 0;
+    }
+    int64_t nanosPerBuffer = (kNanosPerSecond * mLastBackgroundSize) / getSampleRate();
+    const int64_t margin = 200 * kNanosPerMicrosecond; // arbitrary delay so we wake up just after
+    return mBackgroundRanAtNanoseconds + nanosPerBuffer + margin;
+}
+
+// TODO: This method should return a tuple of Result,int32_t where the
```

Comment:
There's a few instances where returning a tuple would certainly be easier to understand, rather than having: 

positive int32_t = some interesting value for the caller
negative int32_t = error which must be cast back to a Result type

Any thoughts?

Reply:
Right, the current way is more appropriate for C. E.g. for sizes, there is a standard type is GNU C library called 'ssize_t' which either contains a positive size or negative error code.

In C++, I would prefer a dedicated type that can be checked as a boolean in a logical expression, and then either the value or an error code can be extracted from it. Here is a good example: https://www.kdab.com/tuple-pair-cpp-apis/  Look for "class value_or_error" in the article. Note that they use std::variant in the implementation, which is from c++17, but we don't actually need it, can just have two fields instead.

Reply:
Hmmm. Maybe. The way that value_or_error template override * seems odd. You use *result to get the value and result.error() to get the error.

Reply:
I implemented this. But I also notice that many oboe functions return a positive value or a negative error. Changing all of them would be disruptive.